Implement EIP 4200 (Static relative jumps)

jochem-brouwer opened the pull request:
Implements [EIP 4200](https://eips.ethereum.org/EIPS/eip-4200), at version https://github.com/ethereum/EIPs/pull/6100

This PR also includes;

- Fix for EIP3670 where it is not possible to run EOF code in a CREATE environment (so CREATE tx, CREATE opcode, CREATE2 opcode)

TODOs

- [ ] Write tests
- [ ] Clarify some stuff in the EIP (noted in this PR)


## Test Cases

### Validation

#### Valid cases

- ~~`RJUMP`/`RJUMPI`/`RJUMPV` with `JUMPDEST` as target~~
    - ~~`relative_offset` is positive/negative/`0`~~
- ~~`RJUMP`/`RJUMPI`/`RJUMPV` with instruction other than `JUMPDEST` as target~~
    - ~~`relative_offset` is positive/negative/`0`~~
- `RJUMPV` with various valid table sizes from 1 to 255

#### Invalid cases

- `RJUMP`/`RJUMPI`/`RJUMPV` with truncated immediate
- `RJUMP`/`RJUMPI`/`RJUMPV` as a final instruction in code section
- `RJUMP`/`RJUMPI`/`RJUMPV` target outside of code section bounds
- `RJUMP`/`RJUMPI`/`RJUMPV` target push data
- `RJUMP`/`RJUMPI`/`RJUMPV` target another `RJUMP`/`RJUMPI`/`RJUMPV` immediate argument
- `RJUMPV` with table size 0

### Execution

- `RJUMP`/`RJUMPI`/`RJUMPV` in legacy code aborts execution
- ~~`RJUMP`~~
    - ~~`relative_offset` is positive/negative/`0`~~
- ~~`RJUMPI`~~
    - ~~`relative_offset` is positive/negative/`0`~~
        - ~~`condition` equals `0`~~
        - ~~`condition` does not equal `0` ~~
- ~~`RJUMPV 1 relative_offset`~~
    - ~~`case` equals `0`~~
    - ~~`case` does not equal `0`~~
- `RJUMPV` with table containing positive, negative, `0` offsets
    - `case` equals `0`
    - `case` does not equal `0` 
    - `case` outside of table bounds (`case >= count`, fallback case)
    - `case` > 255

holgerd77 commented on packages/evm/src/opcodes/functions.ts:
```diff
@@ -760,43 +761,91 @@ export const handlers: Map<number, OpHandler> = new Map([
-      runState.programCounter = Number(dest)
+        const dest = runState.returnStack.pop()
+        runState.programCounter = Number(dest)
+      } else if (common.isActivatedEIP(4200)) {
```
Oh, we've got a double opcode occupancy here? 🤔

jochem-brouwer replied:
Yup. But the other EIP (Subroutines) seems to me a kind of predecessor to this EIP, which is also why most likely the same opcodes are choosen.

holgerd77 commented on packages/common/src/eips/4200.json:
```diff
@@ -0,0 +1,26 @@
+  "url": "https://eips.ethereum.org/EIPS/eip-4200",
+  "status": "Review",
+  "minimumHardfork": "berlin",
+  "requiredEIPs": [3540],
```
This should point to 3670 I think, since the EIP expands on the 3670 rule set. 3540 requirement is then implicitly added as well (along the 3670 activation).

jochem-brouwer replied:
You are right, great catch, it is also in the EIP!

holgerd77 commented on packages/evm/src/eof.ts:
```diff
@@ -62,33 +64,97 @@ export const codeAnalysis = (container: Buffer) => {
 }
 
-export const validOpcodes = (code: Buffer) => {
+export const validOpcodes = (code: Buffer, common: Common) => {
```
Hmm, I am not sure if this method might have made it in the half-official part of the API and people use this in the wild. 🤔 Maybe we should think of an alternative way to add a second required parameter.

Maybe we take an optional parameter `common` and then do `common?.isActivatedEIP(4200)`? Might have less overall side effects.

jochem-brouwer replied:
I honestly don't think people are using this out in the wild, but I might be wrong. I can indeed make it optional.

jochem-brouwer replied:
Made it optional (this also fixes our API tests)

holgerd77 replied:
Haha, are we counting as "the wild" as well? 😜

jochem-brouwer commented:
I just realized we have a problem here. If you activate EIP 4200, then on legacy code these newly introduced opcodes are INVALID, but in EOF they are valid opcodes.

However, in EVMs `step` event, we will report `RJUMP` opcodes, also in legacy. This means we should somehow alter our design such that we have an EVM running into either EOF mode or in legacy mode, such that `step` event reports the right opcodes in both cases.

I will not address this in this PR but it is a structural change we should do rather soon.

jochem-brouwer commented:
If I try to edit the initial post it for some reason deletes a lot of the text, so will post here: CI will most likely pass, so this is now ready for a first round of reviews.

Also note: this PR introduces running EOF code inside CREATE environments. Currently, if you try to run EOF code in a CREATE environment (so either a contract-creating transaction, or any of the CREATE opcodes) then it tries to run opcode 0xEF (INVALID) and immediately goes out-of-gas.

holgerd77 commented on packages/common/src/eips/4200.json:
```diff
@@ -0,0 +1,26 @@
+  "url": "https://eips.ethereum.org/EIPS/eip-4200",
+  "status": "Review",
+  "minimumHardfork": "berlin",
+  "requiredEIPs": [3540, 3670],
```
Short nit: since we have a transitive relation here I would (generally) have a tendency here to keep this to the minimal set and just add 3670. So 3540 will be automatically checked along 3670 activation, and if we add both here we unnecessarily double the 3540 check.

jochem-brouwer replied:
I understand your point, but I wanted to be consistent with the EIP specification. At the time that I wrote this code, it still requires both of those (yes, this is a transitive / circular relation). I would say, in the first iteration of implementing Big EOF, lets try to be as close to the EIP spec, and we can optimize/update this later (for now, we should ensure that it works, it looks clean-ish, and it passes the tests (either generated by us, or what would be more lovely - ethereum/tests unit tests))

holgerd77 commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+      ],
+      // RJUMPI
+      // RJUMPI, jump to JUMPDEST
+      [getEOFCode('5B60015D' + getInt16Str(1) + '5B5B00'), 'RJUMPI to JUMPDEST, +1'],
```
(maybe one can just also make a short comment to make this more explicit and readable, also regarding that there is also a `6000` comment)

jochem-brouwer replied:
Yep, the PUSH 1 is there to ensure that RJUMPI is jumping to the target (and not just ignoring it, i.e. if stackmost top item is 0)

holgerd77 reviewed (commented):
Still on it  but will just submit a first round of review comments to not run the risk of loosing on browser reload or something.

holgerd77 commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+    const codes = [
+      ['5C00', 'truncated RJUMP'],
+      ['5D00', 'truncated RJUMPI'],
+      ['5E0100', 'truncated RJUMPV'],
```
Can we add a `5E02000000` case here?

jochem-brouwer replied:
Good point :+1:

holgerd77 commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+      ['5C' + getInt16Str(2) + '5B0060005B' + getRJUMPVCode([-9]), 'RJUMPV as final instruction'],
+      ['5C' + getInt16Str(1) + '00', 'RJUMP targets code outside container'],
+      ['60015D' + getInt16Str(1) + '00', 'RJUMPI targets code outside container'],
+      ['6000' + getRJUMPVCode([1]) + '00', 'RJUMPV targets code outside container'],
```
Can we add negative target-out-of-code cases here as well?

jochem-brouwer replied:
Also a good point, not sure why I missed this because I think this was in the OG test specs :thinking:

holgerd77 reviewed (commented):
Some things to address and/or discuss, but this looks already pretty good!!! 🤩

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+  const common = new Common({
+    chain: Chain.Mainnet,
+    hardfork: Hardfork.London,
+    eips: [3540, 3670, 4200],
```
I created this PR before I found out about https://github.com/ethereum/EIPs/pull/6156, I think we can all merge it together and check for 3540 to see if EOF contracts are enabled or not (and not do the whole "is EIP `X` activated" thing)

holgerd77 replied:
Ah, wait for it, sorry I was confused since this is only the *activation* of all these EIPs and nothing else.

Let's otherwise please handle the EIPs separately for now and not artificially "merge" into 3540 (if that's what you proposed), that would for sure confuse people and we would deviate from what the EIP contains.

I guess if we want a unifying solution here we would need to carefully discuss (since lightclient mentioned there will likely be non combinatory one-EIP-rule-them-all EIP number). We could "invent" our number and then allow for both (activating separately or by this invented number, hmm. Not sure if this is so optimal.

jochem-brouwer replied:
To clarify, I did not mean to merge all of them into 3540, but at this point (as of https://github.com/ethereum/EIPs/blob/4e5cd9fe6dd168165a7a5cfef68ead439ef27f17/ this version of the EIPs), then if you activate 3540 you will activate all big EOF EIPs (if you do not do this then common will throw) - so internally in our code we can just check if 3540 is activated, then we implicitly also know that 4200 is activated.

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+}
+
+function getEOFCode(code: string) {
+  const magic = '0xEF000101'
```
I need to update this one, thanks :)

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+        'RJUMPV to ADDRESS, -8',
+      ],
+      [getEOFCode('6001' + getRJUMPVCode([1]) + '00FE'), 'RJUMPV with case > count'],
+      [getEOFCode('61FFFF' + getRJUMPVCode([1]) + '00FE'), 'RJUMPV with case > 255'],
```
You are right regarding checking the test cases, I do check if the last opcode is STOP but this does not guarantee we jump to the right JUMPDEST/ADDRESS (should probably SSTORE after and then do a STOP)

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+        'RJUMPI to ADDRESS, -7',
+      ],
+      // RJUMPI: 0 is pushed on stack, so code will just continue after RJUMPI
+      [getEOFCode('60005D' + getInt16Str(1) + '006000FE'), 'RJUMPI, 0 is pushed on stack'],
```
You are right regarding `FE` (INVALID), I am not sure why I added the PUSH 0 though because this makes no sense (I should update these test cases with notes)

jochem-brouwer replied:
Actually, it does, it needs to put 0 on the stack (because RJUMPI pops one stack item). If I do not push0, then the stack is empty and thus stack underflows.

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+  for (const i of int16list) {
+    str += getInt16Str(i)
+  }
+  return str
```
I am not sure if I get your question right, but the check that we jump to the right relative offset is done in the randomly generated tests (this explicitly checks that the storage slot at key zero is the right jump address)

holgerd77 replied:
No, that was not yet my question 😋: the format for `RJUMV` in the EIP is `RJUMPV count relative_offset+` and `count` needs to match the number of `realtive_offset` entries (that's correct, right?).

On the validation side though it is not stated in the EIP that the above matching is actually tested along code validation, the EIP only states `Validation algorithm of [EIP-3670](https://eips.ethereum.org/EIPS/eip-3670) is extended to verify that count >= 1.`.

Should this exact matching also be part of the EIP? 🤔

jochem-brouwer replied:
You should read it as: `count` should not be 0. The check that there exists enough code after the `RJUMPV` opcode (so `2 * count` bytes) is checked in the EOF validation. If there is a wrong count and any of these "jumptable fields" contains an invalid opcode, then EOF validation will also fail.

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+      ],
+      // To ADDRESS
+      [getEOFCode('600030' + getRJUMPVCode([1]) + '303000'), 'RJUMPV to ADDRESS, +1'],
+      [getEOFCode('600030' + getRJUMPVCode([0]) + '303000'), 'RJUMPV to ADDRESS, 0'],
```
You are right, this makes the test suite more complete. What would it make even more complete is to SSTORE the jump address, so we ensure we actually jump to the right position (now we only check for last opcode = STOP)

jochem-brouwer commented on packages/evm/src/evm.ts:
```diff
@@ -551,7 +570,8 @@ export class EVM implements EVMInterface {
           if (
             !EOF.validOpcodes(
-              result.returnValue.slice(codeStart, codeStart + eof1CodeAnalysisResults.code)
+              result.returnValue.slice(codeStart, codeStart + eof1CodeAnalysisResults.code),
```
I think you mean here that we have somewhat hard-coded either legacy contracts or EOF contracts? I very much agree that this should be loosened, because at this point I am not very happy with it (and that is an understatement :joy: )

jochem-brouwer commented on packages/evm/src/opcodes/codes.ts:
```diff
@@ -291,6 +291,14 @@ const eipOpcodes: { eip: number; opcodes: OpcodeEntry }[] = [
+    opcodes: {
+      0x5c: { name: 'RJUMP', isAsync: false, dynamicGas: false },
+      0x5d: { name: 'RJUMPI', isAsync: false, dynamicGas: false },
+      0x5e: { name: 'RJUMPV', isAsync: false, dynamicGas: false },
```
`RJUMPV` is "basically" an upgraded version of `RJUMPI`, the first byte after `RJUMPV` is whatever count the table size is, now you check the stack: is this stack item lower than the table size? In that case, lookup the `int16` of that table and then jump to it, otherwise, just jump right after the table. So, dynamic gas is not applicable here. (I had a lot of trouble understanding this opcode)

holgerd77 replied:
Yeah, but isn't the *size* of this table dynamic and you could therefore bloat the state with a big table size? 🤔

jochem-brouwer replied:
No, the size of this table is determined at code deployment. The `count` is thus hardcoded in the code itself. The `case` is an item which is popped from stack.

Note that the costs for these "big tables" is contained into deployment costs of the code: for each byte you deploy, you have to pay gas (I think it is something like 32 gas per byte or something)

jochem-brouwer commented on packages/vm/test/api/EIPs/eip-4200-static-relative.spec.ts:
```diff
@@ -0,0 +1,222 @@
+      ['5C' + getInt16Str(1) + '00', 'RJUMP targets code outside container'],
+      ['60015D' + getInt16Str(1) + '00', 'RJUMPI targets code outside container'],
+      ['6000' + getRJUMPVCode([1]) + '00', 'RJUMPV targets code outside container'],
+      ['5C' + getInt16Str(1) + '600000', 'RJUMP targets PUSH data'],
```
`RJUMP` targeting `PUSH` data? This is checked by seeing if we jump to an immediate; https://github.com/ethereumjs/ethereumjs-monorepo/blob/11a9e967a431496695ae897a0fcc78ab3febe40f/packages/evm/src/eof.ts#L164

jochem-brouwer commented on packages/common/src/eips/4200.json:
```diff
@@ -0,0 +1,26 @@
+  "comment": "Static relative jumps",
+  "url": "https://eips.ethereum.org/EIPS/eip-4200",
+  "status": "Review",
+  "minimumHardfork": "berlin",
```
Hmm I think you have a point here somewhat, we have not really defined what the "minimum hardfork" actually means. Technically you could also implement EOF (in current spec) right after Chainstart - except that now opcodes like `CREATE2` are not available...

jochem-brouwer replied:
But note: if we have minimum hardfork merge, then this means that non-PoS chains cannot support this EIP.

holgerd77 replied:
Yeah, totally ok with this, very much no strong opinion here.

jochem-brouwer commented:
Hi @holgerd77 I have addressed your points. 

I will re-mark this PR as draft, I want to implement all EIPs into https://github.com/ethereumjs/ethereumjs-monorepo/pull/2453. If you are OK with all changes, please close this PR :smile: (not closing it myself so we do not lost your valuable comments! :smile: )

jochem-brouwer commented:
I have changed the target branch, if you look into the changes then you can see exactly what I changed :) (If undesired change back target branch to master)

jochem-brouwer commented:
Wait why did it merge!?

jochem-brouwer commented:
I am not even at pc right now..,?

jochem-brouwer commented:
Oh wow it auto merges if you do it locally... and then push

holgerd77 commented:
> Hi @holgerd77 I have addressed your points.
> 
> I will re-mark this PR as draft, I want to implement all EIPs into #2453. If you are OK with all changes, please close this PR 😄 (not closing it myself so we do not lost your valuable comments! 😄 )

Thanks for adressing! 🙂 

Did this "PR merged accidentally" issue resolved and if so: how? 😋 (generally good to have a note on this here)

jochem-brouwer commented:
If you `git merge` locally and then push it, and the origin branch is of a PR (i.e. this one) and the target branch is also that of the target branch (i.e. https://github.com/ethereumjs/ethereumjs-monorepo/pull/2453 ) then this will also show as "merged" on github (and this actually makes a lot of sense if you think about it!)

holgerd77 commented:
> If you `git merge` locally and then push it, and the origin branch is of a PR (i.e. this one) and the target branch is also that of the target branch (i.e. #2453 ) then this will also show as "merged" on github (and this actually makes a lot of sense if you think about it!)

Ah ok, thanks for the clarification, didn't notice that this *hasn't* been merged into `master` but "just" the broader EOF branch. 🙂
